applications/tari_watcher/src/config.rs

```diff
@@ -0,0 +1,180 @@
+    pub executable_path: Option<PathBuf>,
+    pub compile: Option<CompileConfig>,
+    pub env: Vec<(String, String)>,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct CompileConfig {
+    pub working_dir: Option<PathBuf>,
+    pub package_name: String,
+    pub target_dir: Option<PathBuf>,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct InstanceConfig {
+    pub name: String,
+    pub instance_type: InstanceType,
+    pub num_instances: u32,
+    #[serde(alias = "extra_args")]
+    pub settings: HashMap<String, String>,
+}
+
+impl InstanceConfig {
+    pub fn new(instance_type: InstanceType) -> Self {
+        Self {
+            name: instance_type.to_string(),
+            instance_type,
+            num_instances: 1,
+            settings: HashMap::new(),
+        }
+    }
+
+    pub fn with_name<S: Into<String>>(mut self, name: S) -> Self {
+        self.name = name.into();
+        self
+    }
+
+    pub fn with_num_instances(mut self, num_instances: u32) -> Self {
+        self.num_instances = num_instances;
+        self
+    }
+}
+
+pub fn get_base_config(cli: &Cli) -> anyhow::Result<Config> {
+    let executables = vec![
+        ExecutableConfig {
+            instance_type: InstanceType::TariValidatorNode,
+            executable_path: Some("target/release/minotari_node".into()),
+            compile: Some(CompileConfig {
+                working_dir: Some("../tari".into()),
+                package_name: "minotari_node".to_string(),
+                target_dir: None,
+            }),
+            env: vec![],
+        },
+        ExecutableConfig {
+            instance_type: InstanceType::MinoTariConsoleWallet,
+            executable_path: Some("target/release/minotari_wallet".into()),
+            compile: Some(CompileConfig {
+                working_dir: Some("../tari".into()),
+                package_name: "minotari_wallet".to_string(),
+                target_dir: None,
+            }),
+            env: vec![],
+        },
+    ];
+    let instances = [
+        InstanceConfig::new(InstanceType::TariValidatorNode)
+            .with_name("tari_validator_node")
+            .with_num_instances(1),
+        InstanceConfig::new(InstanceType::MinoTariConsoleWallet)
+            .with_name("minotari_wallet")
+            .with_num_instances(1),
+    ];
+
+    let base_dir = cli
+        .common
+        .base_dir
+        .clone()
+        .or_else(|| {
+            cli.get_config_path()
+                .canonicalize()
+                .ok()
+                .and_then(|p| p.parent().map(|p| p.to_path_buf()))
+        })
+        .unwrap_or_else(|| std::env::current_dir().unwrap());
+
+    Ok(Config {
+        auto_register: true,
+        base_node_grpc_address: "localhost:18142".to_string(),
+        base_wallet_grpc_address: "localhost:18143".to_string(),
+        sidechain_id: None,
+        vn_registration_file: base_dir.join("vn_registration.toml"),
```

Comment:
This refers to a [JSON file](https://github.com/tari-project/tari-dan/blob/6c1dbc666a6a9a5b6b1524cf3d6460ffb2c67221/applications/tari_validator_node/src/bootstrap.rs#L357) that the validator node creates on startup that holds information required to register the validator node
```suggestion
        vn_registration_file: base_dir.join("registration.json"),
```